resetResponseHandler does not set an onIgnored handler

The resetResponseHandler does not set an onIgnored callback. If that handler is called for a bolt message of type "ignored" then you get a SIGSEV as follows:


```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5ec720]


goroutine 1 [running]:
github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*messageQueue).receive(0xc0002020f8, {0x71e750, 0xc0000b6000})
        /home/trule/go/pkg/mod/github.com/neo4j/neo4j-go-driver/v5@v5.8.1/neo4j/internal/bolt/message_queue.go:166 +0x200
github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*bolt4).ForceReset(0xc000202000, {0x71e750, 0xc0000b6000})
        /home/trule/go/pkg/mod/github.com/neo4j/neo4j-go-driver/v5@v5.8.1/neo4j/internal/bolt/bolt4.go:821 +0x1bc
github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*bolt4).Reset(0xc000000140?, {0x71e750?, 0xc0000b6000?})
        /home/trule/go/pkg/mod/github.com/neo4j/neo4j-go-driver/v5@v5.8.1/neo4j/internal/bolt/bolt4.go:803 +0x7d
github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/pool.(*Pool).Return(0xc000000140, {0x71e750, 0xc0000b6000}, {0x720e70?, 0xc000202000?})
        /home/trule/go/pkg/mod/github.com/neo4j/neo4j-go-driver/v5@v5.8.1/neo4j/internal/pool/pool.go:450 +0x2b2
github.com/neo4j/neo4j-go-driver/v5/neo4j.(*sessionWithContext).executeTransactionFunction.func1()
        /home/trule/go/pkg/mod/github.com/neo4j/neo4j-go-driver/v5@v5.8.1/neo4j/session_with_context.go:436 +0x36
```

The callback is thus (see `onIgnored: nil`):

```
(dlv) print callback
github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.responseHandler {
    onSuccess: github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*bolt4).resetResponseHandler.func1, 
    onRecord: nil, 
    onFailure: github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*bolt4).resetResponseHandler.func2, 
    onUnknown: github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/bolt.(*bolt4).resetResponseHandler.func3, 
    onIgnored: nil
}
```

Adding `onIgnored: onIgnoredNoOp,` to the resetResponseHandler() would likely fix the fault.

https://github.com/neo4j/neo4j-go-driver/blob/0eccfa3db544f6639151bf013cb08a3dde97fc70/neo4j/internal/bolt/bolt4.go#LL1078C1-L1091C2


The scenario behind this is fairly simple (and probably stupid) ... however, I'm new to Go ... based of the example, I basically run this function twice with the same parameters, the second time fails.

```
func addNodeFunc(ctx context.Context, labels []string, name string) neo4j.ManagedTransactionWork {
	return func(tx neo4j.ManagedTransaction) (any, error) {
		var b strings.Builder
		b.WriteString("CREATE (n:" + strings.Join(labels, ":") +" {name: $name}) ")
		b.WriteString("RETURN id(n) ")
		query := b.String()
		result, err := tx.Run(ctx, query, map[string]any{"name": name})
		if err != nil {
			return nil, err
		}
		return result.Consume(ctx)
	}
}
```

@timrulebosch thanks for the report!
I have not been to reproduce the issue just yet.
What server version are you running?

Here are the steps I followed:

```shell
docker run --rm \
    --env NEO4J_AUTH='neo4j/letmein!' \
    --env NEO4J_ACCEPT_LICENSE_AGREEMENT=yes \
    --publish=7687:7687 \
    --health-cmd "cypher-shell -u neo4j -p 'letmein!' 'RETURN 1'" \
    --health-interval 5s \
    --health-timeout 5s \
    --health-retries 5 \
    neo4j:4.4-enterprise
```

And I ran the following program (with an invalid label):

```go
package main

import (
	"context"
	"fmt"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
	"strings"
)

func main() {
	ctx := context.Background()
	driver, err := neo4j.NewDriverWithContext("neo4j://localhost", neo4j.BasicAuth("neo4j", "letmein!", ""))
	if err != nil {
		panic(err)
	}
	defer dealWithIt(ctx, driver)
	session := driver.NewSession(ctx, neo4j.SessionConfig{})
	defer dealWithIt(ctx, session)
	for i := 0; i < 5; i++ {
		fmt.Printf("Call #%d\n", i+1)
		summary, err := session.ExecuteWrite(ctx, addNodeFunc(ctx, []string{"F o o", "Bar"}, "Jane Doe"))
		if err != nil {
			panic(err)
		}
		fmt.Printf("Node(s) created: %d\n", summary.(neo4j.ResultSummary).Counters().NodesCreated())
	}
}

func addNodeFunc(ctx context.Context, labels []string, name string) neo4j.ManagedTransactionWork {
	return func(tx neo4j.ManagedTransaction) (any, error) {
		var b strings.Builder
		b.WriteString("CREATE (n:" + strings.Join(labels, ":") + " {name: $name}) ")
		b.WriteString("RETURN id(n) ")
		query := b.String()
		result, err := tx.Run(ctx, query, map[string]any{"name": name})
		if err != nil {
			return nil, err
		}
		return result.Consume(ctx)
	}
}

func dealWithIt(ctx context.Context, closer interface{ Close(context.Context) error }) {
	if err := closer.Close(ctx); err != nil {
		panic(err)
	}
}

```

which gives me:
```go
Call #1
panic: Neo4jError: Neo.ClientError.Statement.SyntaxError (Invalid input 'o': expected ")", "WHERE", "{" or a parameter (line 1, column 13 (offset: 12))
"CREATE (n:F o o:Bar {name: $name}) RETURN id(n)"
             ^)

goroutine 1 [running]:
main.main()
```

Would you be able to update the program and make it crash as in the initial report?
Thanks a lot!

--- a/neo4j/internal/bolt/message_queue.go
+++ b/neo4j/internal/bolt/message_queue.go
@@ -159,32 +159,32 @@ func (q *messageQueue) receive(ctx context.Context) error {
 	case *db.Record:
 		onRecord := handler.onRecord
 		if onRecord == nil {
-			return errors.New("this response handler does not support RECORD")
+			return errors.New("protocol violation: the server sent an unexpected RECORD response")
 		}
 		onRecord(message)
 	case *success:
 		onSuccess := handler.onSuccess
 		if onSuccess == nil {
-			return errors.New("this response handler does not support SUCCESS")
+			return errors.New("protocol violation: the server sent an unexpected SUCCESS response")
 		}
 		onSuccess(message)
 	case *db.Neo4jError:
 		onFailure := handler.onFailure
 		if onFailure == nil {
-			return errors.New("this response handler does not support FAILURE")
+			return errors.New("protocol violation: the server sent an unexpected FAILURE response")
 		}
 		onFailure(ctx, message)
 		return message
 	case *ignored:
 		onIgnored := handler.onIgnored
 		if onIgnored == nil {
-			return errors.New("this response handler does not support IGNORED")
+			return errors.New("protocol violation: the server sent an unexpected IGNORED response")
 		}
 		onIgnored(message)
 	default:
 		onUnknown := handler.onUnknown
 		if onUnknown == nil {
-			return fmt.Errorf("this response handler does not support unknown %v", message)
+			return fmt.Errorf("protocol violation: the server sent an unknown %v response", message)
 		}
 		onUnknown(message)
 	}

--- a/neo4j/internal/bolt/message_queue.go
+++ b/neo4j/internal/bolt/message_queue.go
@@ -23,6 +23,7 @@ import (
 	"container/list"
 	"context"
 	"errors"
+	"fmt"
 	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
 	"github.com/neo4j/neo4j-go-driver/v5/neo4j/log"
 	"net"
@@ -153,19 +154,39 @@ func (q *messageQueue) receive(ctx context.Context) error {
 	if q.handlers.Len() == 0 {
 		return errors.New("no more response callback to apply")
 	}
-	callback := q.pop()
+	handler := q.pop()
 	switch message := res.(type) {
 	case *db.Record:
-		callback.onRecord(message)
+		onRecord := handler.onRecord
+		if onRecord == nil {
+			return errors.New("this response handler does not support RECORD")
+		}
+		onRecord(message)
 	case *success:
-		callback.onSuccess(message)
+		onSuccess := handler.onSuccess
+		if onSuccess == nil {
+			return errors.New("this response handler does not support SUCCESS")
+		}
+		onSuccess(message)
 	case *db.Neo4jError:
-		callback.onFailure(ctx, message)
+		onFailure := handler.onFailure
+		if onFailure == nil {
+			return errors.New("this response handler does not support FAILURE")
+		}
+		onFailure(ctx, message)
 		return message
 	case *ignored:
-		callback.onIgnored(message)
+		onIgnored := handler.onIgnored
+		if onIgnored == nil {
+			return errors.New("this response handler does not support IGNORED")
+		}
+		onIgnored(message)
 	default:
-		callback.onUnknown(message)
+		onUnknown := handler.onUnknown
+		if onUnknown == nil {
+			return fmt.Errorf("this response handler does not support unknown %v", message)
+		}
+		onUnknown(message)
 	}
 	return nil
 }
@@ -195,8 +216,8 @@ func (q *messageQueue) receiveMsg(ctx context.Context) any {
 	return msg
 }
 
-func (q *messageQueue) enqueueCallback(callbacks responseHandler) {
-	q.handlers.PushBack(callbacks)
+func (q *messageQueue) enqueueCallback(handler responseHandler) {
+	q.handlers.PushBack(handler)
 }
 
 func (q *messageQueue) setLogId(logId string) {

--- a/neo4j/internal/bolt/message_queue_test.go
+++ b/neo4j/internal/bolt/message_queue_test.go
@@ -245,7 +245,7 @@ func TestMessageQueue(outer *testing.T) {
 						writer.appendX(msgRecord, []any{})
 						writer.send(ctx, server)
 					},
-					expectedErrorMsg: "this response handler does not support RECORD",
+					expectedErrorMsg: "protocol violation: the server sent an unexpected RECORD response",
 				},
 				{
 					description: "SUCCESS",
@@ -253,7 +253,7 @@ func TestMessageQueue(outer *testing.T) {
 						writer.appendX(msgSuccess, map[string]any{})
 						writer.send(ctx, server)
 					},
-					expectedErrorMsg: "this response handler does not support SUCCESS",
+					expectedErrorMsg: "protocol violation: the server sent an unexpected SUCCESS response",
 				},
 				{
 					description: "FAILURE",
@@ -261,7 +261,7 @@ func TestMessageQueue(outer *testing.T) {
 						writer.appendX(msgFailure, map[string]any{})
 						writer.send(ctx, server)
 					},
-					expectedErrorMsg: "this response handler does not support FAILURE",
+					expectedErrorMsg: "protocol violation: the server sent an unexpected FAILURE response",
 				},
 				{
 					description: "IGNORED",
@@ -269,7 +269,7 @@ func TestMessageQueue(outer *testing.T) {
 						writer.appendX(msgIgnored)
 						writer.send(ctx, server)
 					},
-					expectedErrorMsg: "this response handler does not support IGNORED",
+					expectedErrorMsg: "protocol violation: the server sent an unexpected IGNORED response",
 				},
 			}
 
@@ -281,7 +281,7 @@ func TestMessageQueue(outer *testing.T) {
 					go func() {
 						err := queue.receive(ctx)
 
-						AssertErrorMessageContains(inner, err, test.expectedErrorMsg)
+						AssertErrorMessageContains(t, err, test.expectedErrorMsg)
 						done <- struct{}{}
 					}()
 

--- a/neo4j/internal/bolt/message_queue_test.go
+++ b/neo4j/internal/bolt/message_queue_test.go
@@ -22,6 +22,7 @@ package bolt
 import (
 	"bytes"
 	"context"
+	"fmt"
 	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
 	. "github.com/neo4j/neo4j-go-driver/v5/neo4j/internal/testutil"
 	"net"
@@ -227,7 +228,70 @@ func TestMessageQueue(outer *testing.T) {
 			writer.send(ctx, server)
 			<-done
 		})
+
+		inner.Run("returns error when nil callback called for", func(inner *testing.T) {
+			inner.Parallel()
+
+			type testCase struct {
+				description      string
+				writerWork       func(*outgoing)
+				expectedErrorMsg string
+			}
+
+			testCases := []testCase{
+				{
+					description: "RECORD",
+					writerWork: func(o *outgoing) {
+						writer.appendX(msgRecord, []any{})
+						writer.send(ctx, server)
+					},
+					expectedErrorMsg: "this response handler does not support RECORD",
+				},
+				{
+					description: "SUCCESS",
+					writerWork: func(o *outgoing) {
+						writer.appendX(msgSuccess, map[string]any{})
+						writer.send(ctx, server)
+					},
+					expectedErrorMsg: "this response handler does not support SUCCESS",
+				},
+				{
+					description: "FAILURE",
+					writerWork: func(o *outgoing) {
+						writer.appendX(msgFailure, map[string]any{})
+						writer.send(ctx, server)
+					},
+					expectedErrorMsg: "this response handler does not support FAILURE",
+				},
+				{
+					description: "IGNORED",
+					writerWork: func(o *outgoing) {
+						writer.appendX(msgIgnored)
+						writer.send(ctx, server)
+					},
+					expectedErrorMsg: "this response handler does not support IGNORED",
+				},
+			}
+
+			for _, test := range testCases {
+				inner.Run(fmt.Sprintf("%s response", test.description), func(t *testing.T) {
+					done := make(chan any)
+					queue.enqueueCallback(responseHandler{})
+
+					go func() {
+						err := queue.receive(ctx)
+
+						AssertErrorMessageContains(inner, err, test.expectedErrorMsg)
+						done <- struct{}{}
+					}()
+
+					test.writerWork(writer)
+					<-done
+				})
+			}
+		})
 	})
+
 }
 
 func assertEqualResponseHandlers(t *testing.T, handler1, handler2 responseHandler) {

